cmd/tag-set.go

```diff
@@ -155,11 +178,34 @@ func mainSetTag(cliCtx *cli.Context) error {
 
 	if timeRef.IsZero() && !withVersions {
 		setTags(ctx, clnt, versionID, tags)
+		for content := range clnt.List(ctx, ListOptions{TimeRef: timeRef}) {
```

Comment:
@shtripat I think a simpler change is:
- Add `&& !recursive` in L179
- Add 'ListOptions{Recursive: recursive}' in L194

setTagsForSubDirs() does not seem to be needed in my opinion, especially after adding --recursive flag

Reply:
Ok, let me change it and verify once.

Reply:
With below code
```
	if timeRef.IsZero() && !withVersions && !recursive {
		setTags(ctx, clnt, versionID, tags)
	} else {
		for content := range clnt.List(ctx, ListOptions{TimeRef: timeRef, WithOlderVersions: withVersions, Recursive: true}) {
			if content.Err != nil {
				fatalIf(content.Err.Trace(), "Unable to list target "+targetURL)
			}
			setTags(ctx, clnt, content.VersionID, tags)
		}
```
The problem persists and results are as below
```
./mc tag set myminio/test2 "key=val-bkt-test2-rec" --recursive
Tags set for http://localhost:9000/test2.
Tags set for http://localhost:9000/test2.
Tags set for http://localhost:9000/test2.
Tags set for http://localhost:9000/test2.

./mc tag set myminio/test3 "key=val-bkt-test3-rec" --recursive
Tags set for http://localhost:9000/test3.
Tags set for http://localhost:9000/test3.
Tags set for http://localhost:9000/test3.
```
Actually for bucket only its looping and setting the tags again and again, for no of times of the count of subdir objects. That's the reason the function
```
func setTagsForSubDirs(ctx context.Context, cliCtx *cli.Context, content *ClientContent, targetURL string, tags string) {
	bucket, _ := url2BucketAndObject(&content.URL)
	targetURL = targetURL + string(content.URL.Separator) +
		strings.TrimPrefix(
			content.URL.String(),
			content.URL.Scheme+"://"+content.URL.Host+string(content.URL.Separator)+bucket+string(content.URL.Separator))
	clnt, err := newClient(targetURL)
	fatalIf(err.Trace(cliCtx.Args()...), "Unable to initialize target "+targetURL)
	setTags(ctx, clnt, content.VersionID, tags)
}
```
recreates the client with proper targetURL of the subdir objects. Initial client is created with targetURL as say `myminio/test2` whereas in case of `recursive=True` and dubdirs available we should create client with targetPath as say `myminio/test2/1/issue`, `myminio/test2/2/issue` and `myminio/test2/3/issue`.

May be function name `setTagsForSubDirs` is not too intuitive :(

Reply:
@shtripat my point is just to simplify this further, and I think setTagsForSubDirs can be avoided.

My suggestion is to implement something similar to setRetentionSingle() and calls listing only if --recursive or --versions is passed

You can DM anyway for further information.

Reply:
Also `--versions` would make sure tag gets set for all the versions. Whereas `--recursive` for all all subdirs. If both passed then for all versions of all subdirs.

Reply:
@vadmeste as suggested changed the logic inline with `setRetentionSingle` as works as expected. Also added `--recursive` to tags listing command as well.
I suspect `mc tag remove` may also need some fixes but we can take up in separate PR